L2 event pulse quorum confirmations

jagerman opened the pull request:
This implements a confirmation mechanism into the pulse quorum to enhance the security of pulse blocks without introducing desyncing of the Oxen chain to achieve consensus.

The very high level overview is that this changes the validation of L2 confirmations so that, rather than every node validating an event, only the pulse quorum votes on events, and only after sufficient votes in favour of an event are attained does a Eth event get applied to the oxen chain.

In particular this means that Oxen nodes do *not* require instant consensus with L2 events to sync the *Oxen* chain; instead they trust everything the pulse quorum says so that all Oxen nodes are always fully in agreement as to the current state of the Session service node network, and move validation/acceptance/rejection of L2 state changes into the pulse quorum.

The general design implemented here is explained in the `docs/sent-l2-confirmation/sent-confirmations.md` document added in this PR.

Note that this PR breaks the serialization format of HF20/21, and will requiring restarting devnet and stagenet chains when it comes into effect.

Draft PR for now because although this implementation is believed feature complete, it still requires testing, review, and discussion (and some conflict resolution).

jagerman committed:
SENT L2 tx confirmation overview doc & examples

jagerman committed:
Remove non-operator contributor minimum staking requirements for ETH regs

Enforcement of minimum stakes now happens in the multi-contributor
contract.  There are, however, cases in which the old strictly ordered
Oxen rules might not be satisfied (e.g. if a large contributor withdraws
from an in-progress open stake), but a full set of contributions can
*always* be rearranged to satisfy the rule.  But not enforcing it
Oxen-side we can save a little trouble of needing the multi contributor
contract to worry about sorting the contributor list before submitting
it to the SN contract.

jagerman committed:
Restore pre-ETH_BLS registration signatures

The PR adding the BLS registration command removed the old registration
hash, but it is still very much needed to be able to sync the (pre-ETH)
chain.

jagerman committed:
Add map serialization support

Maps serialization works with the generic container + pair serialization
code with some small tweaks.

The main tweak is that we need to be able to specify the (otherwise
deduced) Container::value_type because the one for pair has a read-only
key, but we need the mutable key alternative for insertion.

Also includes a fix for the `.insert` code path not actually working.

jagerman committed:
Remove devnet hardfork

This was added for localdev, but in current WIP PRs localdev now has its
own hardfork table and this isn't needed anymore.

jagerman committed:
remove OXEN_LOG_ENABLED macro

jagerman committed:
Print Synced messages in globallogcat

So that they show up at the default log level.

jagerman committed:
De-inline globallogcat

jagerman committed:
Properly apply --l2-provider-check intervals

Checks to compare all L2 providers were firing at every (10s) L2 update,
instead of the intended default 2-minute interval, because the
next-check timestamp wasn't being set.

This also fixes the command-line options to override the default not
being set.

jagerman committed:
Move process_ethereum_* definitions together

(No change).

jagerman committed:
Simplify

jagerman committed:
Fix wrong output iterator on write_varint concept

We only write chars, not Ts, into the output iterator.  It was working
because, for integer type T, you can technically output a T into a char,
but working doesn't mean correct.

Also fixes some minor doc errors (grammar + EVARINT instead of
EVARIANT).

jagerman committed:
rpc: only put block_header or block_headers in output if requested

If you only use it in single hash or multi-hash mode then you should
only see the single header or array of headers output.  Currently in
single hash mode it was always including an empty block_headers array
which was confusing.

Also shortens the `get_block_header_by_hash` name to `gbh` because the
previous name was too verbose (and cumbersome to type) for a variable
with such a short scope.

jagerman committed:
Remove miner_tx from HF21+ blocks

In HF19 we still use it (because batched rewards go into it), but in
HF21 it should never be present and thus can be dropped from
serialization; this converts it to an optional and makes it never
present in HF21+.

Also here are some closely related changes:

- add a transaction_prefix.is_miner_tx() rather than having the
  has-1-vin-and-it-is-a-txin_gen condition sprinkled all over the
  codebase.

- Add a `block.get_height()` that understands when to get the height out
  of the miner_tx's txin_gen and when to get it out of the block member,
  and use it rather than various places that expected the txin_gen to be
  present.

- Fixes block boost::archive serialization (used in some blockchain
  utilities): the `ver < 19` and whatnot code was *completely* broken
  (because `ver` is a serialization, which was not version bumped).
  This bumps it to ver 1, and fixes it to support nullopt miner_tx
  values.

jagerman committed:
Avoid double compilation of t_cryptonote_protocol_handler

These .inl files are so much yuck.

jagerman committed:
De-clutter `core`

- De-clutter pointless wrappers from core into various subsystems.
  Anything that can access core can already access those subsystems
  directly and it doesn't make sense to put tons of methods through core
  rather than the relevant subsystem.  (Doing so seems to be a holdover
  of Monero reluctance to ever properly refactor their codebase).

- Rename references to "blockchain storage" to just "blockchain" because
  it hasn't been "blockchain storage" in the better part of a decade.

- Make subsystem member references available directory out of core and
  Blockchain, rather than needing to go through a reference accesssor.
  For instance, `m_core.get_blockchain_storage()` is now just
  `m_core.blockchain`, `m_core.get_pool()` is now `m_core.mempool`, and
  so on.  Since we exposed all of these reference directly through the
  accessors anyway, keeping them a private member reference rather than
  just exposing them directly made no sense.

- Remove `get_` prefix from the subsystems that aren't a simple member
  variable (such as `get_omq()` -> `omq()`, `get_bls_signer()` ->
  `bls_signer()`, `blockchain.get_db()` -> `blockchain.db()`), to be
  closer to the now-exposed member references.

- De-virtualize `core` because the only reason it was virtual was
  because the miner implementation was lazy.  Replace the miner
  interface pointer with the three needed callbacks instead.

jagerman committed:
fix compat with fmt 9

jagerman committed:
Drop blockchain_blackball

We've never used it and (especially with SENT), never will, and it makes
some assumptions about miner_tx existence that haven't been valid for a
while on Oxen.

jagerman committed:
Compat hack for current devnet

jagerman committed:
Promote registration rejection logs to warnings

jagerman committed:
ETH HF registrations: enforce full stake

In theory the smart contract should already not allow this, but double
checking on the oxen side seems sensible.

jagerman committed:
Log reasons in decomm/dereg log messages

jagerman committed:
Remove 9 year old comments about 10 year old code

jagerman committed:
Methodize block_has_pulse_components

Just makes it a bit nicer to use.

jagerman committed:
Fix block add error message

- should be error, not info
- wasn't including the actual exception message in the log string format

jagerman committed:
friendly_duration: Don't print fractions for whole seconds

Currently the "status" command-line command prints something like:

    Height: xxx, ..., uptime 13.000s

even though the value can only ever be a whole number.  This change
suppresses the ".000" from friendly_duration when the given duration is
a whole seconds value.

jagerman committed:
Add design docs about block reward determination

jagerman committed:
formatting

jagerman committed:
Confirmations, part 1: reward consensus & structure changes

Various changes needed to work towards L2 confirmation consensus,
including an implementation of the block reward consensus logic.

- Rearrange block data to support HF21 confirmations

  The current HF19 `block` layout is suboptimal: blocks (but not the
  block_header) has a redundant `height` field (redundant because HF19
  blocks *always* have a miner_tx also containing the height), has a
  mislabelled "service_node_winner_key" (which actually contains the
  pulse round producer key and, on backup rounds, disagrees with the
  actual service node winner in the miner_tx tx_extra), and has a reward
  field in the block rather than block_header.

  Fields in `block` are *not* signed by the block hash and thus are
  susceptible to undetectable forgeries at the P2P layer, where you
  could (for example) alter the reward value before handing the block
  off to a syncing node to cause them to unrecoverably stall syncing
  once the error propagates into the reward database.

  This redoes the fields for HF19:

  - height is moved to block_header, and becomes `height_`, with a
    `get_height()` method added to actually return the height from
    either the miner_tx txin_gen property (≤HF20) or else from the
    height_ field (HF21+).
  - reward moves to the block_header.
  - l2_height was in block rather than block_header and so wasn't
    signed.
  - service_node_winner_key is renamed to `oxen10_pulse_producer` and
    will be droppable once everyone has upgraded to Oxen 11.  (The
    upgrade mechanics for dropping it are described in blockchain.cpp).

- Add new block header fields for HF20/21:

  - `reward` (moved from block) is repurposed to return the *actual*
    reward under HF21, which is not (necessarily) identical to the
    current l2 reward, but rather is based on consensus rules around
    recent l2 reward values (see added design docs from the previous
    commit for details).
  - `sn_winner_tail` replaces the miner_tx tx_extra service_node_winner
    and includes just the last 4 bytes (32 bits) of the winner pubkey
    for effective enough verification while keeping the size down.  This
    field was never crucial as we always check it against what the value
    *should* be when checking a block (and always use the proper value),
    so only checking its last 4 bytes is enough to rule out the sync
    errors that would be caught by this.
  - `l2_height` (moved from block)
  - `l2_reward` is a clampeed smart contract pool reward rate.  The
    clamp here is only allowing it to change very slowly from the
    previous block's `l2_reward`; see added design docs from previous
    commit for more details.

- Tightened up the rule for HF19/20 block reward: previously the block's
  reward value was only required to be <= the required value, but now it
  must be exactly equal.  (I ran this code through a full mainnet sync
  to verify that there are no HF19 blocks violating this equality).

- Replace unchecked `service_node_winner_key` for determining tx fee
  recipient in db_sqlite with the proper pulse leader as required by the
  current service node state.  (This makes the service_node_winner_key
  AKA oxen10_pulse_producer only used to propagate the info to Oxen 10
  nodes that still expect it in the serialized block, while Oxen 11
  otherwise ignores it).

- l2_reward is now partially consensus determined: it is permitted only
  tiny increases/decreases from one block to the next.

- reward is now entirely consensus determined (by recent l2_reward
  values).

- Replaced block setup with new l2_reward/reward values (as per new
  design) for HF21.

- Added constants for l2_reward and reward consensus, both for initial
  FH21 blocks and once the blocks are flowing.

- Remove min L2 height from L2 mempool admission; we now just take a
  maximum and can admit whatever L2 state changes are hanging around in
  the mempool from any (not-too-recent) height.

- Added some FIXMEs about issues in the ethereum removal processing that
  need to be addressed.

- Added state_t methods to retrieve the pulse producer & quorum that
  produced the current state_t, as this is needed in multiple places.
  (There was a `get_block_producer` already in the header, but it didn't
  actually have an implementation).

- Renamed `get_block_leader()` to `get_next_blocker_leader()` to be
  clearer what it is actually returning, especially w.r.t. to the
  methods added above.

- Add a cache of the state_t's next block leader value as it is
  non-trivial to retrieve and we call it multiple times on the way
  through adding/syncing a block.

- Expanded on the descriptions of different HF miner tx vout
  compositions

- Stop hiding the miner_tx_hash in HF19/20 RPC code even if it has no
  vouts.  It is still a transaction, is still in the database, and still
  has a tx hash.

jagerman committed:
Serialization map fixes; vector<bool> support; drop pair

- Add support for serializing a vector<bool>; this gets serialized as a
  varint length followed by packed (8 bit per byte) bool values.
- Remove std::pair serialization: it was a stupid implementation
  that encoded as `2 || FIRST || SECOND` for who knows what reason, but
  also isn't used anywhere.
- Fix map/unordered_map serialization that didn't work through generic
  container serialization because of the constness of the map key.

jagerman committed:
Pulse quorum L2 event confirmations

This implements a confirmation mechanism into the pulse quorum to
enhance the security of pulse blocks without introducing desyncing of
the Oxen chain to achieve consensus.

The general design implemented here is explained in the
`docs/sent-l2-confirmation/sent-confirmations.md` document (and so see
that for details rather than this commit message).

Other smaller changes included here that intersect (but aren't directly
part of the confirmation mechanism):

- Combine the virtually identical `tx_extra_ethereum_new_service_node`
  and `NewServiceNodeTx` instead a single 'eth::event::NewServiceNode`,
  and likewise for removal requests and removals.
- Update the naming of L2 events to match the more informative names in
  https://github.com/oxen-io/eth-sn-contracts/pull/58
- Increase L2 refresh times to once per minute.
- Remove useless pass-through wrapper functions for txpool operations
  from blockchain.cpp, and just have tx_pool call directly into the db
  via blockchain's `db()` reference.
- Remove l2 monotonic height checks from consensus consideration; they
  are dangerous in that a malicious quorum could stall the chain by
  putting an excessive value that honest nodes couldn't follow.
- Various related code cleanup in passing.
- renamed some functions such as `get_locked_key_image_unlock_height`
  that have become rather misnamed by the eth transition.
- Simplify tx_pool sorting which was, for completely inexplicable
  reasons, a pair<tuple<A,B,C>, D> rather than a tuple<A,B,C,D>.
- Refomatting

Doy-lee commented on src/cryptonote_basic/cryptonote_basic.h:
```diff
@@ -480,13 +546,50 @@ void serialize_value(Archive& ar, block& b) {
+            // Serialize a 0 height, even if the value we have isn't 0 (which it might not be if we
+            // added this block to our DB while running Oxen 10).
+            [[maybe_unused]] uint64_t no_height = 0;
+            field_varint(ar, "height", no_height);
```
When a Oxen 11+ node syncs blocks from HF19 from an Oxen 11+ node it's going to set `rewards` to 0, then for example wouldn't this branch fail in `Blockchain::validate_miner_transaction`?

```cpp
  } else if (version == hf::hf19_reward_batching) {
      if (b.reward != reward_parts.miner_fee + reward_parts.service_node_total) {
          log::error(
                  log::Cat("verify"),
                  "block reward for batching is incorrect: {} != {} ({} SN + {} tx fees)",
                  print_money(b.reward),
                  print_money(reward_parts.miner_fee + reward_parts.service_node_total),
                  print_money(reward_parts.service_node_total),
                  print_money(reward_parts.miner_fee));
          return false;
      }
}
```


`pulse_producer` on the otherhand is fine because we pull out the value from `tx_extra`. I think it's even possible to get rid of it now in the `SUPPORT_OXEN10_INTEROP` branch though I'm guessing that's probably going to have a cyclical include problem.

```cpp
if constexpr (SUPPORT_OXEN10_INTEROP) {
    // ...
    if (Archive::is_deserializer && b._height == 0) {
        // ...
    } else {
        auto pulse_producer = cryptonote::get_service_node_winner_from_tx_extra(b.miner_tx->extra);
        field(ar, "pulse_producer", pulse_producer);
        // ...
    }
}
```

jagerman replied:
Good catch!

(The code was from before I realized that we really still want the reward alongside the block info, and I updated the *comment* to say that we still used it but not the actual code.)

It's fixed now so that just height/pulse producer are in the compat code, but reward is outside it.

jagerman replied:
> pulse_producer on the otherhand is fine because we pull out the value from tx_extra. I think it's even possible to get rid of it now in the SUPPORT_OXEN10_INTEROP branch though I'm guessing that's probably going to have a cyclical include problem.

I looked at this, and I don't think it is because current Oxen 10 code uses that value from the serialized block in some places -- so dropping it now would break Oxen 10's ability to sync from an Oxen 11 node in the period where we have both on the network during the last days of HF19.

Doy-lee replied:
> > pulse_producer on the otherhand is fine because we pull out the value from tx_extra. I think it's even possible to get rid of it now in the SUPPORT_OXEN10_INTEROP branch though I'm guessing that's probably going to have a cyclical include problem.
> 
> I looked at this, and I don't think it is because current Oxen 10 code uses that value from the serialized block in some places -- so dropping it now would break Oxen 10's ability to sync from an Oxen 11 node in the period where we have both on the network during the last days of HF19.

In this case it wouldn't be dropping it because in the serialisation phase an Oxen 11 node can retrieve the pulse producer value from a HF19 block (by extracting it from the TX extra) and serialise it so that it's always present on the wire when its being synced for Oxen 10 nodes. 

Though it's pretty minor .. because we can just wait until we drop the code when we're all on Oxen 11.

jagerman committed:
Add l2_height to L2 events

This solves the problem that events with identical values (such as
re-submitting a registration long after the previous one has expired but
with the same operator, fee, etc.) would result in duplicate tx_hashes
and thus not be allowed on the Oxen chain.

jagerman committed:
Fix: always serialize reward in HF19

This was broken from a half-edited implementation in the previous
commits: we still use and want the serialized `reward` in HF19, it's
just height and winner that should be under the Oxen 10 compat mode.
(We still have it in HF20, too, but in HF20+ it's serialized in
`block_header` instead of `block`).

This fixes it by moving the reward serialization outside the compat
branches.

jagerman committed:
Fix debug build

The serializer hates being given a const object, so don't constify this
reference.

Doy-lee commented on src/l2_tracker/events.h:
```diff
@@ -16,7 +15,15 @@ using namespace std::literals;
 
-struct L2StateChange {};
+struct L2StateChange {
+    uint64_t l2_height = 0;
```
Thoughts on having a domain-separation tag like we do in BLS schemes and other chains do differentiate the hashes per network type?

jagerman replied:
Can you elaborate on what this would look like?

Doy-lee replied:
For example the BLS tag is `H("<command>" || chain_id || contract_address)`

It's probably just enough to do 

```
H(l2_height || network_chain_id || tx_label || tx_blob)`

# With tx_label being "ServiceNodeRemovalRequest", "ServiceNodeRemoval" or "NewServiceNode".
```

It's not too annoying to smuggle in the network chain ID into the tx_extra here as there's only one place to construct these L2 state changes.

jagerman replied:
I'd added the chainId into the L2StateChange base class now, and added it to serialization (commit: `Add chainId to events; check chainId on startup`, 2482b347b1c9b770cd31a957843e16aaf60ffa2f).  (And along the way also added verification of the chainId of all L2 providers on startup).

The separate hashing isn't necessary here because it's all encompassed by the overall tx hash that applies on the serialized tx extra data, which (now) contains chain id, and there's already a unique tx type value in the serialized transaction for each event type; so a tx hash already depends on each of l2_height, chain_id, event type, and other event parameters via that generic transaction hashing.

Doy-lee commented on src/cryptonote_core/tx_pool.cpp:
```diff
@@ -235,34 +237,15 @@ bool tx_memory_pool::have_duplicated_non_standard_tx(
-        if (!cryptonote::get_field_from_tx_extra(tx.extra, data)) {
+    } else if (is_l2_event_tx(tx.type)) {
+        std::string fail;
+        auto event = eth::extract_event(hard_fork_version, tx, fail);
```
Are we missing a check to ensure that the L2 event doesn't already exist in the txpool? The other branches do this kind of business

```cpp
        std::vector<transaction> pool_txs;
        get_transactions(pool_txs);
        for (const transaction& pool_tx : pool_txs) {
            if (pool_tx.type != txtype::state_change)
                continue;
```

Doy-lee replied:
Also this check might need to be more robust with "double spends", or for instance yesterday's situation by using for example `Blockchain::handle_block_to_main_chain`'s check:

```cpp
        if (m_db->tx_exists(tx_id)) {
            log::info(
                    logcat,
                    fg(fmt::terminal_color::red),
                    "Block with id: {} attempting to add transaction already in blockchain with "
                    "id: {}",
                    id,
                    tx_id);
            return false;
        }
```

Seems it's possible for a duplicate TX hash to be put into the TX pool (one that exists in the DB but not in the tx pool). Once it's in it stalls the network until someone coordinates a flush of the pool? Or it expires (can't remember if that happens or not) But if we block dupe TX's at the TX pool level this should be impossible to do at all.

I can't tell if we check for dupe TX hashes for other transactions, I see we have dupe key-image checks but our custom TX types without key-images in general could benefit from also having this check hoisted up earlier in the function to eliminate that issue.

jagerman replied:
Yes, I'll add it.  I thought that it would just fail if it already exists, but it looks like it removes it and then reinserts it which seems unnecessary.

jagerman replied:
Added.

It doesn't look like we ever check for identical tx hashes, but I don't think we need to because the other checks being done are stronger: e.g. obligation quorum state change txes could have different subsets of the validator signatures for the same state change (and thus different hashes, even though it is the exact same change); and a regular tx could be spending the same inputs but include different random components.  In each case an *actually* identical hash is just a subset of the possible duplicate transactions.

Doy-lee commented on src/cryptonote_core/service_node_list.h:
```diff
@@ -776,8 +868,13 @@ class service_node_list {
+        // blockchaindb-global-transaction-index => confirmation metadata of unconfirmed L2 state
+        // changes.  This is an *ordered* map because confirmation votes in a pulse block depend on
+        // the order of txes in here.
+        std::map<crypto::hash, unconfirmed_l2_tx> unconfirmed_l2_txes;
```
Is the blockchaindb-global-transaction-index the 32 byte hash, because the comment on `l2_votes` says it's a monotonic tx index?

```
    // vector of L2 state changes for L2 state change transactions that are still unconfirmed in
    // this block.  true = vote for confirmation, false = vote for rejection.  Votes are sorted in
    // blockchain order of unconfirmed transactions (i.e. by the monotonic internal transaction
    // index).
    std::vector<bool> l2_votes;
```

~~If it were the hash there'd be an issue with the mapping of block votes to unconfirmed txs since they'd get re-sorted depending on the new L2 tx hashes input into the SNL~~

--

Oh I think I've got it, the votes are only valid for that very height, they potentially get invalidated but refreshed and remapping 1:1 with unconfirmed_l2_txes on block creation.

jagerman replied:
> Is the blockchaindb-global-transaction-index the 32 byte hash, because the comment on l2_votes says it's a monotonic tx index?

Oops, leftover comment: I was trying to use the global tx index, but there is apparently no way to get a tx out of the database by global index, which is why I switched to hash.

It *does* change the sort order because now they are sorted by hash, but that's okay because, as you said, it only matters within that single block.

jagerman replied:
Pushed a fix for the misleading comment.

Doy-lee commented on src/cryptonote_core/service_node_list.cpp:
```diff
@@ -2855,14 +3001,75 @@ void service_node_list::state_t::update_from_block(
+                        throw oxen::traced<std::runtime_error>{
+                                "Internal error: did not find state change tx data in blockchain database: {}"_format(
+                                        fail)};
+                    need_swarm_update += std::visit(
```
Does a removal request (e.g. start 15 day or 30 (can't remember) timer) warrant a swarm update?

jagerman replied:
Good catch; I've fixed the remreq overload to return false.

Doy-lee commented on docs/sent-l2-confirmation/sent-confirmations.md:
```diff
@@ -0,0 +1,357 @@
+  that it learns about would lag the rest of the Oxen network, causing it to fall behind.
+
+- A malicious L2 provider who provides a significant amount of the Oxen network could provide false
+  information regarding registration activity.
```
A L2 RPC provider could inject fake transactions and get it relayed across to all the other nodes. The other nodes will then provide confirmation on these transactions causing havoc when they weren't actually in the L2. 

This seems like the most feasible attack. The P2P code is pretty complicated, I saw that when we add a L2 transaction we mark it as do-not-relay. Is there a distinction in the P2P layer from TXs that are sent via relay with zero-confirmations and TXs that are sent as part of syncing the block (e.g. somewhat confirmed)? 

The idea being we reject L2 TXs on the P2P layer if they aren't in a block because we're only meant to observe these TXs from L2 whilst our RPC node still has the data. If we go out of the available range of blocks in the RPC provider then we just have to trust the node you're syncing from.

jagerman replied:
> This seems like the most feasible attack. The P2P code is pretty complicated, I saw that when we add a L2 transaction we mark it as do-not-relay. Is there a distinction in the P2P layer from TXs that are sent via relay with zero-confirmations and TXs that are sent as part of syncing the block (e.g. somewhat confirmed)?

The do-not-relay code is not actually necessary anymore because when the pulse quorum signs off on a transaction, it requires that any L2 state changes are known in the L2 tracker, not the mempool (and likewise for subsequent pulse quora casting confirmation votes).

Doy-lee replied:
Confirmed not a problem. This is the code in `create_block_template_internal`

```
b.l2_votes = service_node_list.l2_pending_state_votes();
```

Which applies votes based on what's in the L2 tracker, _not_ the tx pool

jagerman committed:
No swarm update needed for removal reqs

jagerman committed:
Fix l2_votes comment about sort order

The global tx output order didn't play nicely with what blockchain_db
exposes, so the actual code as committed sorts by hash instead.

Doy-lee commented:
This generally looks pretty sound to me, I'll have a quick re-review again when it's rebased

jagerman committed:
Block syncing optimizations

- Don't compute the PoW hash for HF16+ blocks: the value we compute is
  only used as a cache, but for pulse blocks we *never use the value*.
  Cutting out this useless calculation speeds up testnet syncing (on my
  machine) from under 200 blocks/s to over 600 blocks/s.

  If we *do* need the pow hash (e.g. for mined fallbacks) then it
  already gets computed on demand if not in the cache.

- Remove "amount" handling from block syncing code.  Non-zero amounts
  are an early Monero pre-ringct thing that has never applied to Oxen as
  Oxen has always been ringct.

- inline and remove output_scan_worker (because it never gets used in a
  worker since we never had more than one event value).

jagerman committed:
Add block sync speed to Synced messages

Before:

    Synced 413816/584100  (70%, 170284 left)

After:

    Synced 413816/584100  (70%, 170284 left)  (616 blocks/sec)

The speed is quite useful to see to get an idea of how long syncing will
take, but is only shown if you put oxend into info+ verbosity, along
with a bunch of other less useful info.  This adds the speed to the
message always (but keeps the other stuff at info+ or debug+ verbosity).

jagerman committed:
Fix wrong pulse quorum calculations

Recent commits in this PR broke the proper pulse quorum calculations;
this fixes them:

- Add block_leader to state_t.  We *need* it to be able to compute the
  next block leader.  Previously we would grab it out of the miner_tx's
  tx_extra when needed, but as we don't have that anymore, we now need
  to store it somewhere.
- Pulse quorum calculation was being broken by the height and hash
  having already changed before we computed it.
- There was some confusion between the *current* pulse quorum that made
  the block/state, and the *next* pulse quorum for the next block that
  is resolved here.

jagerman committed:
Revert vector -> set misoptimization

This ended up making things go slower rather than faster when processing
blocks with lots of inputs.

jagerman committed:
Allow shutting down during subsystem rescan

Currently oxend's subsystem rescan is not abortable, and so if you try
killing or restarting oxend, it simply won't acknoledge the signal.
On service nodes, this often results in systemd timing out the stop and
restoring to a SIGKILL, which is not a nice way to go (and can cause
lmdb corruption or bloating if done at the wrong time).

This commit weaves an atomic<bool> through the potentially slow parts of
core and blockchain initialization so that when we *do* want to stop
(from Ctrl-C, or a regular TERM signal) we gracefully fail and exit
rather than needing a `kill -9` to actually stop.

jagerman committed:
Optimization: skip verification during SN state rescan

When we rescan the service node state from the blockchain (which is
sometimes necessary after a crash, or after certain pop_block patterns)
we do an unnecessary full pulse signature verification of the block, but
this is unnecessary on this path: the fact that we added the block to
the LMDB at some point means it already passed the required checks.

This roughly doubles the SN state rescan speed (because of no longer
needing to extract and verify signatures on each pulse block).

This also lowers the number of blocks we process at once to 100 instead
of 500 as we sometimes have reports that 1) it can respond (to a
shutdown) faster, and 2) on overloaded servers processing 500 blocks can
take so long that it triggers the systemd watchdog (because our ping of
systemd is itself in this loop).  The reduction to 100 didn't appear to
affect performance at all on my (unloaded) system so seems reasonable to
do.

jagerman committed:
Add chainId to events; check chainId on startup

- Adds chainId to eth::event objects so that (chainid, height,
  parameters) that make a unique tx hash.
- Check chainId of all providers at startup and fail if any are wrong.
- Add `--l2-skip-chainid` argument to be able to skip the check
- Update ethyl submodule for getAllChainIds() support
- Add the Arbitrum mainnet chain id into network_config so that you can
  start mainnet syncing with a mainnet Arb L2 provider configured.
- Formatting and small cleanups.

Doy-lee commented on src/l2_tracker/events.h:
```diff
@@ -16,13 +16,21 @@ using namespace std::literals;
+            chain_id{chain_id}, l2_height{l2_height} {}
+
+    template <class Archive>
+    void serialize_base(Archive& ar, uint8_t version) {
```
Version needs to passed in as a reference `uint8_t &version` so that calling code can branch on the version if it changes to conditionally deserialise the binary stream.

```
void serialize_base(Archive& ar, uint8_t version) {
```

Though I don't see a situation where the base `L2StateChange` would change between different ethereum events that we can just inline the version.

```cpp
        uint8_t version = 0;
        field_varint(ar, "version", version);
        field_varint(ar, "chain_id", chain_id);
        field_varint(ar, "l2_height", l2_height);
```

I'd also just include a version in the derived class as well to track changes in those structs independently from the base class. It'll only cost an additional byte to store the version for each TX extra.

jagerman replied:
In retrospect trying to split the serialization into the base class wasn't worth the complexity, so I've just moved the two fields into each derived class and dropped the base class serialization code entirely.

Doy-lee commented on src/l2_tracker/events.h:
```diff
@@ -16,13 +16,21 @@ using namespace std::literals;
 
 struct L2StateChange {
-    uint64_t l2_height = 0;
+    uint64_t chain_id;
```
It's probably not a problem but we will repeat chain IDs for our test networks because we will re-deploy onto Sepolia Arbitrum, but those are only for our test networks. So be it.

jagerman replied:
That shouldn't be a problem because we still have the l2_height here which will still differ across re-deployments.

Doy-lee commented on src/cryptonote_core/blockchain.cpp:
```diff
@@ -1774,11 +1848,10 @@ bool Blockchain::create_block_template_internal(
-        b.prev_id = get_tail_id();
         median_weight = m_current_block_cumul_weight_limit / 2;
         diffic = get_difficulty_for_next_block(!info.is_miner);
         already_generated_coins = m_db->get_block_already_generated_coins(height - 1);
```
This has to be

```cpp
already_generated_coins = m_db->get_block_already_generated_coins(height);
```

To get the cumulative generated coins (since we use `get_tail_id` which gives the height of the last block committed to the chain)

Doy-lee replied:
On second thought instead of this suggestion, we have to amend this line here

```cpp
        std::tie(height, b.prev_id) = get_tail_id();
```

To convert the height into blockchain height

```cpp
        std::tie(height, b.prev_id) = get_tail_id();
        height += 1;
```

Because the height value returned from create block template and all other code that uses it, expects it to be the height of the blockchain (e.g. the height of the next block to be produced).

jagerman replied:
Good catch; fixed.

jagerman committed:
Un-split base field serialization

Trying to put some of the serialization in the base class just ended up
making it more complicated without much real gain, and will cause a lot
of complications in the future if the version needs to be increased.

jagerman committed:
Fix height calculation in block template creation

jagerman committed:
Reject identical L2 events in the mempool

jagerman committed:
Move network default subdir into network_config

jagerman committed:
Version bump; reboot/redir devnet & stagenet

- 11.0.3, along with requiring >=11.0.3 for uptime proofs.
- Reboot devnet & stagenet networks with new network id & new default
  data directory (devnet2, stagenet2, respectively).
- Update stable version uptime proof requirements (dev had fallen behind
  stable).
- Redeploy devnet DEVSENT contracts for latest contract code.

jagerman committed:
fix broken start_mining help output

jagerman committed:
bump ethyl to latest

jagerman committed:
Fix mempool selection/drop penalty zone under SENT

It was trying to apply Oxen reward values for SENT blocks during mempool
calculation (in order to compute when adding a tx starts reducing the
total reward), but that won't work in SENT because there *are* no Oxen
reward values under SENT to reduce.

Instead this updates the mempool selection algorithm to simply make the
300kB previously "soft" block limit a hard limit starting at HF21,
dropping support for the (rarely used) dynamic mempool size and penalty
zone.

jagerman committed:
Fix Sepolia errors fetching recent reward heights

On Sepolia nets, we sometimes get an error trying to retrieve reward
rates, particularly at startup when we are performing initial fetches.

It turns out this is because Arbitrum Sepolia often goes a bit slower
than 250ms (currently average is more like 360ms) as 250ms is only a
minimum block time, but Arb will sometimes slow down a bit if there are
no blocks when the sequencer makes the next block.

But compounding this is the fact that most Arb nodes only keep 128
blocks + blocks up to 30m old, and so requesting the reward at a height
>30m old will error, which can happen in the current implementation when
the L2 block rate slows down.

Even on mainnet with its very-close-to-250ms block time average the
warning would trigger *occassionally* because if you are starting up at
current L2 height of 1002399 then the heights you need rewards for are
1000000, 997600, 995200, but 995200 is 7199 ago and so is right at the
30min deletion threshold.

This includes a couple fixes to address it:

- Only fetch current and last applicable block reward.  We were
  previously also fetching the second-last one, but with the last one
  being at least 10 minutes old, this is unnecessary as we should never
  need it.  (And so for mainnet, in the worst case example above, we
  would only ever fetch back to 997600, which will typically be only 20
  minutes ago).
- Lower the reward interval on testnet/devnet/stagenet to 600 blocks
  instead of mainnet's 2400 blocks.

With the two changes above, we should be fine for a sustained average L2
block rate (over the last half hour) of up to around 1.5s on Sepolia,
and up to 375ms on mainnet Arbitrum.

It's also worth noting that hitting these isn't fatal: if the L2 chain
stalls for some reason there is an additional buffer provided by Oxen
nodes that have been running for since before the stall: they will hold
onto cached recent values and can still proceed (only recently restarted
ones would have difficulty validating).

Also removes the L2_BLOCK_TIME because it really isn't quite correct
(and was only being used in one place).

jagerman committed:
Fix tx tree hash when a block has no txes

Such a tree hash is now simply defined as the null hash.

jagerman committed:
Various fixes for creating a BLS_ETH block

- `_height` was not being set on the BLS_ETH path
- `block`'s copy/move constructor weren't copying/moving the
  oxen10_pulse_producer or tx_eth_count fields, thus resulting in
  garbage values.
- handle_block_to_main_chain was attempting the validate the
  non-existant miner transaction

jagerman committed:
Fix pool duplicate L2 event check

Doy-lee commented on src/cryptonote_core/blockchain.cpp:
```diff
@@ -1550,49 +1573,100 @@ bool Blockchain::validate_miner_transaction(
-    if (version < feature::ETH_BLS) {
-        if (b.reward >
-            reward_parts.base_miner + reward_parts.miner_fee + reward_parts.service_node_total) {
+    } else if (version == hf::hf19_reward_batching) {
```
Are we missing HF21 here? I think so, we should patch this like so:

```cpp
} else if (version >= hf::hf19_reward_batching && hf <= hf::hf20_eth_transition) {
  // pre-existing code
} else { // NOTE: Change to fallback else branch to capture the rest.
}
```

Suggest avoiding the feature flags so that we can visually pattern match the HF numbers are all being handled correctly.

jagerman replied:
It was indeed missing the checks for 20.  The function logic is now shuffle around quite a bit (and renamed to `validate_block_rewards` instead of `validate_miner_transaction`) because the HF21+ checks are pretty much independent of all the miner_tranactions bits needed for <=HF20.

jagerman replied:
> Suggest avoiding the feature flags so that we can visually pattern match the HF numbers are all being handled correctly.

I agree in principle, but the fix in this case ended up splitting the logic into two versions with virtually no overlap: a "eth-bls" block, and a "non-eth-bls" block.

Doy-lee commented on src/cryptonote_core/blockchain.cpp:
```diff
@@ -1823,43 +1886,76 @@ bool Blockchain::create_block_template_internal(
+        // pool reward changes will only show up gradually in l2_reward, but prevents abuse by
+        // faulty/compromised L2 providers or pulse quorums).
+        auto l2r_range = l2_reward_range(height);
+        b.l2_reward = std::clamp(*actual_reward, l2r_range.first, l2r_range.second);
```
We aren't setting `b.rewards` here and I couldn't see any other place where it is being set for HF21 blocks.
```cpp
        b.reward  = eth_consensus_reward(height);
```
I haven't figured out why this isn't triggering the check in `blockchain::validate_miner_transaction` namely this block of code
```cpp
        if (auto expected_reward = eth_consensus_reward(height); b.reward != expected_reward) {
            log::error(
                    log::Cat("verify"),
                    "block reward for height {} is incorrect; block has {} but expected {}",
                    height,
                    print_money(b.reward),
                    print_money(expected_reward));
            return false;
        }
```

jagerman replied:
It didn't get caught because I broke the validation to fix another issue. [28dbb9b](https://github.com/oxen-io/oxen-core/pull/1697/commits/28dbb9b09f214ddc2fa8eb47182b9068c36e92a4) fixes both issues.

jagerman committed:
Allow "none" as mining address

The mining address is useless under SENT and hasn't really been useful
elsewhere since pulse.  This adds "none" as a valid address which
becomes the 0 address to simply discard any (pre-HF21) rewards.

jagerman committed:
Don't log errors for duplicate L2 event txes

These aren't errors; it just means it's in the mempool already (for
example, because we already put it there and then restarted and
refetched events).

jagerman committed:
Replace toxic monstrosity data structure

No,

        std::vector<std::pair<
                std::pair<std::string, crypto::hash>,
                std::vector<std::pair<crypto::hash, std::string>>>>& blocks,

is not a useful data structure.

Once rewritten to not be a dumpster fire, a bug was easy to spot and fix
in here: that we weren't properly recognizing a null miner tx hash.

jagerman committed:
Add backward compat support for `get_accrued_batched_earnings` endpoint

This got renamed and reworked a bit, but letting the old name still work
for now makes life easier for the observer.

jagerman committed:
Expose actual block_size in RPC endpoints

We were setting `block_size` to the `block_weight`, but that was wrong:
block weight is only the sum of tx sizes.  `block_size` is now set to
the total block size (i.e. actual block data + sum of tx sizes).

jagerman committed:
formatting

jagerman committed:
Fix l2_reward inclusion and validation

Fix l2_reward and reward not being set (or validated) properly for
ETH_BLS blocks.

jagerman committed:
Fix l2_votes json serialization

l2_votes was being json serialized as packed values, but that isn't
right for a vector<bool> when going to json: we want to see a json array
of true/false values (both because it's more obvious, but also because
the packed values can't distinguish between false-prefix arrays (i.e.
`[false,true,true]` and `[true,true]` both become [3]; for binary
archives this isn't a problem because those always include the size (and
so we *can* distinguish between the above two cases since they have size
3 and 2, respectively).

jagerman committed:
Prevent already-mined L2 events from entering mempool

Events were only being checked against other mempool txes, but also need
to be checked against the chain so that we neither add, nor try to
include, such txes in new blocks.

jagerman committed:
Fix popped txes in mempool not having l2_height set

This drops the from_l2 option because we don't actually need it: we
always have it in the tx itself and can just extract it and get it that
way, which also fixes the problem when the event is coming from a popped
block.

Doy-lee commented on src/daemon/command_parser_executor.cpp:
```diff
@@ -443,31 +443,27 @@ bool command_parser_executor::print_transaction_pool_stats(const std::vector<std
         return true;
-    }
-    if (info.is_subaddress) {
+    } else if (info.is_subaddress) {
```
Oh this is sneaky. First time I've seen the scoping trick used.

Doy-lee commented on src/rpc/core_rpc_server_commands_defs.h:
```diff
@@ -2127,7 +2127,7 @@ struct GET_SERVICE_NODE_STATUS : NO_ARGS {
 ///    atomic SENT units).
 struct GET_ACCRUED_REWARDS : PUBLIC {
-    static constexpr auto names() { return NAMES("get_accrued_rewards"); }
+    static constexpr auto names() { return NAMES("get_accrued_rewards", "get_accrued_batched_earnings"); }
```
The observer has been updated to use `get_accrued_rewards` I suspect the devnet.observer just needs to pull the latest committed changes on the oxen 11 compat branch.

Doy-lee commented on src/cryptonote_core/tx_pool.cpp:
```diff
@@ -269,6 +269,14 @@ bool tx_memory_pool::have_duplicated_non_standard_tx(
             }
         }
+
+        std::unique_lock b_lock{m_blockchain};
```
Do we know where these TXs are coming from? Could be the P2P layer if it's from a malicious node, but in normal operations ... How does this get triggered?

Doy-lee commented:
Some comments but no changes required. I have a patch for fixing the unit tests that I'll apply on this branch, run the local devnet test and merge if it's all good.

Doy-lee committed:
Update the unit test interfaces to accept block size ptr

Doy-lee commented:
Some of the unit tests are broken but its more to do with stupid mock stuff that's gone out of date. Not important. Merging.
